### PR TITLE
fix: separate out subscription errors from field level resolver generation

### PR DESCRIPTION
**Pull request by danielleadams**
#### Description of changes

This pull requests separates out generation of field-level resolvers from checking for errors on types with subscriptions configured.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by ammarkarachi — APPROVED** [review]

File: packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import { $TSAny, $TSContext, stateManager, exitOnNextTick, FeatureFlags, pathManager } from 'amplify-cli-core';
+import { collectDirectivesByType, readProjectConfiguration } from 'graphql-transformer-core';
+import { printer, prompter } from 'amplify-prompts';
+import { DocumentNode, FieldNode, DirectiveNode, parse } from 'graphql';
+import path from 'path';
+import fs from 'fs-extra';
+
+async function flipNotificationFlag(projectPath: string, flagName: string, value: boolean): Promise<void> {
```
Comment (jhockett) [inline packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts:8]: nit: something like `setNotificationFlag` might be a better name

File: packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  let config = stateManager.getCLIJSON(projectPath, undefined, {
+    throwIfNotExist: false,
+    preserveComments: true,
+  });
+
+  config.features.graphqltransformer[flagName] = value;
+  stateManager.setCLIJSON(projectPath, config);
```
Comment (jhockett) [inline packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts:15]: We should either throw if the file doesn't exist, or we need better handling in the event that config is undefined.

File: packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+export async function notifyFieldAuthSecurityChange(context: $TSContext): Promise<void> {
+  const flagName = 'showfieldauthnotification';
+  const dontShowNotification = !FeatureFlags.getBoolean(`graphqltransformer.${flagName}`);
+
+  if (dontShowNotification) return;
+
+  const projectPath = pathManager.findProjectRoot() ?? process.cwd();
+  const meta = stateManager.getMeta();
```
Comment (jhockett) [inline packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts:26]:
```suggestion
  const meta = stateManager.getMeta(projectPath);
```

File: packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    const subscriptionOff = modelDirective.arguments.some((arg: any) => {
+      const levelFieldOff = arg.value?.fields?.some(({ name, value }) => name.value === 'level' && value.value === 'off');
+      return arg.name.value === 'subscriptions' && levelFieldOff;
+    });
+
+    return subscriptionOff && fieldDirectives.has(typeName);
+  });
+}
+
+export function haveFieldAuthDirectives(doc: DocumentNode): Set<string> {
```
Comment (SwaySway) [inline packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts:78]: nit
```suggestion
export function hasFieldAuthDirectives(doc: DocumentNode): Set<string> {
```

File: packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+  await flipNotificationFlag(projectPath, flagName, false);
+}
+
+export function displayAuthNotification(directiveMap: any, fieldDirectives: Set<string>): boolean {
+  return Object.keys(directiveMap).some((typeName: string) => {
+    const typeObj = directiveMap[typeName];
+    const modelDirective = typeObj.find((dir: DirectiveNode) => dir.name.value === 'model');
+
+    const subscriptionOff = modelDirective.arguments.some((arg: any) => {
```
Comment (SwaySway) [inline packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts:69]: it's possible there might be a schema with no `@model`s
```suggestion
    const subscriptionOff = modelDirective?.arguments?.some((arg: any) => {
```

**Review by SwaySway — APPROVED** [review]
lgtm aside from potential scenario with no `@model` in schema

**Comment by danielleadams** [comment]
@jhockett @SwaySway addressed comments

**Review by jhockett — APPROVED** [review]
LGTM

**Review by undefobj — APPROVED** [review]

File: packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      `This version of Amplify CLI introduces additional security enhancements for your GraphQL API. ` +
+        `The changes are applied automatically with this deployment. This change won't impact your client code. Continue`,
+    );
+
+    if (!continueChange) {
+      await context.usageData.emitSuccess();
+      exitOnNextTick(0);
+    }
+
+    const schemaPath = path.join(apiResourceDir, 'schema.graphql');
```
Comment (lazpavel) [inline packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts:65]: this needs to handle the case for projects with `schema` directory

**Review by lazpavel — CHANGES_REQUESTED** [review]

File: packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts (hunk 6 above)
Comment (lazpavel) [reply packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts:65]: a similar change had to be added for the other notification: https://github.com/aws-amplify/amplify-cli/pull/9670

File: packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
+import { $TSAny, $TSContext, stateManager, exitOnNextTick, FeatureFlags, pathManager } from 'amplify-cli-core';
+import { collectDirectivesByType, readProjectConfiguration } from 'graphql-transformer-core';
+import { printer, prompter } from 'amplify-prompts';
+import { DocumentNode, FieldNode, DirectiveNode, parse } from 'graphql';
+import path from 'path';
+import fs from 'fs-extra';
+
+async function setNotificationFlag(projectPath: string, flagName: string, value: boolean): Promise<void> {
+  let config = stateManager.getCLIJSON(projectPath, undefined, {
```
Comment (lazpavel) [inline packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts:9]: nit: can probably use `FeatureFlags.ensureFeatureFlag('graphqltransformer', flagName);`

**Review by lazpavel — APPROVED** [review]

File: packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts (hunk 6 above)
Comment (danielleadams) [reply packages/amplify-cli/src/extensions/amplify-helpers/auth-notifications.ts]: thanks for pointing this out - fixed.

**Pull request merged**